### PR TITLE
feat: Added Stable Matching Algorithm

### PR DESCRIPTION
**Pull request by BKarthik7**
# feat: Added Stable Matching Algorithm

## Description

This commit introduces the Stable Matching Algorithm, commonly known as the Gale-Shapley Algorithm, to the codebase. It solves the problem of matching two sets (e.g., men and women) based on preference lists, ensuring that the final matching is stable—meaning no two individuals would rather be paired with each other than with their current partners.

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] I ran bellow commands using the latest version of **rust nightly**.
- [x] I ran `cargo clippy --all -- -D warnings` just before my last commit and fixed any issue that was found.
- [x] I ran `cargo fmt` just before my last commit.
- [x] I ran `cargo test` just before my last commit and all tests passed.
- [x] I added my algorithm to the corresponding `mod.rs` file within its own folder, and in any parent folder(s).
- [x] I added my algorithm to `DIRECTORY.md` with the correct link.
- [x] I checked `COUNTRIBUTING.md` and my code follows its guidelines.

### CODE REVIEW

File: src/greedy/stable_matching.rs [hunk 1]
```diff
@@ -0,0 +1,155 @@
+use std::collections::{BTreeMap, VecDeque};
+
+// This function performs the Gale-Shapley stable matching algorithm
+pub fn stable_matching(
+    men_preferences: &BTreeMap<String, Vec<String>>,
+    women_preferences: &BTreeMap<String, Vec<String>>,
```
Comment (vil02) [inline src/greedy/stable_matching.rs:6]: is the `BTreeMap` needed here? Would it be enough to use `HashMap`?

File: src/greedy/stable_matching.rs [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+    // Return the final stable matches as a BTreeMap
+    let mut stable_matches: BTreeMap<String, String> = BTreeMap::new();
+    for (man, woman_option) in man_engaged {
+        if let Some(woman) = woman_option {
+            stable_matches.insert(man, woman);
+        }
+    }
+
+    stable_matches
+}
```
Comment (vil02) [inline src/greedy/stable_matching.rs:80]: This is too long.

File: src/greedy/stable_matching.rs [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+        let matches = stable_matching(&men_preferences, &women_preferences);
+        let expected_matches1 = BTreeMap::from([]);
+        let expected_matches2 = BTreeMap::from([]);
+
+        // Assert that the result matches one of the expected outcomes
+        assert!(
+            matches == expected_matches1 || matches == expected_matches2,
+            "Matching result is not as expected"
+        );
+    }
```
Comment (vil02) [inline src/greedy/stable_matching.rs:154]: Please split this into two.

**Review by vil02 — CHANGES_REQUESTED** [review]
Please add the _missing_ tests. We are aiming to _100%_ coverage.

File: src/greedy/stable_matching.rs (hunk 2 above)
Comment (BKarthik7) [reply src/greedy/stable_matching.rs:80]: @vil02 Didn't understood? removed comments to make it short

Comment (BKarthik7) [reply src/greedy/stable_matching.rs:80]: @vil02 any other ideas to make it short

Comment (vil02) [reply src/greedy/stable_matching.rs]: Probably there is a way to _cut_ this function into smaller ones. Consider this as refactoring.

**Comment by BKarthik7** [comment]
> Please add the _missing_ tests. We are aiming to _100%_ coverage.

done it reached 100% convergence

**Comment by BKarthik7** [comment]
@vil02  If I consider refactoring the code losses its nice readability is it ok?

File: src/greedy/stable_matching.rs (hunk 2 above)
Comment (BKarthik7) [reply src/greedy/stable_matching.rs]: @vil02 If I consider refactoring the code losses its nice readability is it ok?

Comment (BKarthik7) [reply src/greedy/stable_matching.rs]: @vil02 something like use std::collections::{HashMap, VecDeque};

pub fn stable_matching(
    men_preferences: &HashMap<String, Vec<String>>,
    women_preferences: &HashMap<String, Vec<String>>,
) -> HashMap<String, String> {
    let (mut free_men, mut current_partner, mut man_engaged, mut next_proposal) =
        initialize_data_structures(men_preferences, women_preferences);

    while let Some(man) = free_men.pop_front() {
        handle_proposal(
            &man,
            &mut free_men,
            &mut current_partner,
            &mut man_engaged,
            &mut next_proposal,
            men_preferences,
            women_preferences,
        );
    }

    extract_stable_matches(man_engaged)
}

fn initialize_data_structures(
    men_preferences: &HashMap<String, Vec<String>>,
    women_preferences: &HashMap<String, Vec<String>>,
) -> (
    VecDeque<String>,
    HashMap<String, Option<String>>,
    HashMap<String, Option<String>>,
    HashMap<String, usize>,
) {
    let mut free_men: VecDeque<String> = VecDeque::new();
    let mut current_partner: HashMap<String, Option<String>> = HashMap::new();
    let mut man_engaged: HashMap<String, Option<String>> = HashMap::new();
    let mut next_proposal: HashMap<String, usize> = HashMap::new();

    for man in men_preferences.keys() {
        free_men.push_back(man.clone());
        next_proposal.insert(man.clone(), 0);
    }

    for woman in women_preferences.keys() {
        current_partner.insert(woman.clone(), None);
    }

    (free_men, current_partner, man_engaged, next_proposal)
}

fn handle_proposal(
    man: &str,
    free_men: &mut VecDeque<String>,
    current_partner: &mut HashMap<String, Option<String>>,
    man_engaged: &mut HashMap<String, Option<String>>,
    next_proposal: &mut HashMap<String, usize>,
    men_preferences: &HashMap<String, Vec<String>>,
    women_preferences: &HashMap<String, Vec<String>>,
) {
    let man_pref_list = &men_preferences[man];
    let next_woman_idx = *next_proposal.get(man).unwrap();
    let woman = &man_pref_list[next_woman_idx];

    next_proposal.insert(man.to_string(), next_woman_idx + 1);

    if let Some(current_man) = current_partner[woman].clone() {
        if woman_prefers_new_man(woman, man, &current_man, women_preferences) {
            update_engagement(man, woman, current_man, free_men, current_partner, man_engaged);
        } else {
            free_men.push_back(man.to_string());
        }
    } else {
        man_engaged.insert(man.to_string(), Some(woman.to_string()));
        current_partner.insert(woman.to_string(), Some(man.to_string()));
    }
}

fn update_engagement(
    man: &str,
    woman: &str,
    current_man: String,
    free_men: &mut VecDeque<String>,
    current_partner: &mut HashMap<String, Option<String>>,
    man_engaged: &mut HashMap<String, Option<String>>,
) {
    man_engaged.insert(man.to_string(), Some(woman.to_string()));
    current_partner.insert(woman.to_string(), Some(man.to_string()));
    free_men.push_back(current_man);
}

fn woman_prefers_new_man(
    woman: &str,
    man1: &str,
    man2: &str,
    preferences: &HashMap<String, Vec<String>>,
) -> bool {
    let woman_preferences = &preferences[woman];
    woman_preferences.iter().position(|m| m == man1).unwrap()
        < woman_preferences.iter().position(|m| m == man2).unwrap()
}

fn extract_stable_matches(
    man_engaged: HashMap<String, Option<String>>,
) -> HashMap<String, String> {
    let mut stable_matches = HashMap::new();
    for (man, woman_option) in man_engaged {
        if let Some(woman) = woman_option {
            stable_matches.insert(man, woman);
        }
    }
    stable_matches
}

Comment (BKarthik7) [reply src/greedy/stable_matching.rs]: '''use std::collections::{HashMap, VecDeque};

pub fn stable_matching(
    men_preferences: &HashMap<String, Vec<String>>,
    women_preferences: &HashMap<String, Vec<String>>,
) -> HashMap<String, String> {
    let (mut free_men, mut current_partner, mut man_engaged, mut next_proposal) =
        initialize_data_structures(men_preferences, women_preferences);

    while let Some(man) = free_men.pop_front() {
        handle_proposal(
            &man,
            &mut free_men,
            &mut current_partner,
            &mut man_engaged,
            &mut next_proposal,
            men_preferences,
            women_preferences,
        );
    }

    extract_stable_matches(man_engaged)
}

fn initialize_data_structures(
    men_preferences: &HashMap<String, Vec<String>>,
    women_preferences: &HashMap<String, Vec<String>>,
) -> (
    VecDeque<String>,
    HashMap<String, Option<String>>,
    HashMap<String, Option<String>>,
    HashMap<String, usize>,
) {
    let mut free_men: VecDeque<String> = VecDeque::new();
    let mut current_partner: HashMap<String, Option<String>> = HashMap::new();
    let mut man_engaged: HashMap<String, Option<String>> = HashMap::new();
    let mut next_proposal: HashMap<String, usize> = HashMap::new();

    for man in men_preferences.keys() {
        free_men.push_back(man.clone());
        next_proposal.insert(man.clone(), 0);
    }

    for woman in women_preferences.keys() {
        current_partner.insert(woman.clone(), None);
    }

    (free_men, current_partner, man_engaged, next_proposal)
}

fn handle_proposal(
    man: &str,
    free_men: &mut VecDeque<String>,
    current_partner: &mut HashMap<String, Option<String>>,
    man_engaged: &mut HashMap<String, Option<String>>,
    next_proposal: &mut HashMap<String, usize>,
    men_preferences: &HashMap<String, Vec<String>>,
    women_preferences: &HashMap<String, Vec<String>>,
) {
    let man_pref_list = &men_preferences[man];
    let next_woman_idx = *next_proposal.get(man).unwrap();
    let woman = &man_pref_list[next_woman_idx];

    next_proposal.insert(man.to_string(), next_woman_idx + 1);

    if let Some(current_man) = current_partner[woman].clone() {
        if woman_prefers_new_man(woman, man, &current_man, women_preferences) {
            update_engagement(man, woman, current_man, free_men, current_partner, man_engaged);
        } else {
            free_men.push_back(man.to_string());
        }
    } else {
        man_engaged.insert(man.to_string(), Some(woman.to_string()));
        current_partner.insert(woman.to_string(), Some(man.to_string()));
    }
}

fn update_engagement(
    man: &str,
    woman: &str,
    current_man: String,
    free_men: &mut VecDeque<String>,
    current_partner: &mut HashMap<String, Option<String>>,
    man_engaged: &mut HashMap<String, Option<String>>,
) {
    man_engaged.insert(man.to_string(), Some(woman.to_string()));
    current_partner.insert(woman.to_string(), Some(man.to_string()));
    free_men.push_back(current_man);
}

fn woman_prefers_new_man(
    woman: &str,
    man1: &str,
    man2: &str,
    preferences: &HashMap<String, Vec<String>>,
) -> bool {
    let woman_preferences = &preferences[woman];
    woman_preferences.iter().position(|m| m == man1).unwrap()
        < woman_preferences.iter().position(|m| m == man2).unwrap()
}

fn extract_stable_matches(
    man_engaged: HashMap<String, Option<String>>,
) -> HashMap<String, String> {
    let mut stable_matches = HashMap::new();
    for (man, woman_option) in man_engaged {
        if let Some(woman) = woman_option {
            stable_matches.insert(man, woman);
        }
    }
    stable_matches
}'''

Comment (BKarthik7) [reply src/greedy/stable_matching.rs]: @vil can i push this

**Comment by BKarthik7** [comment]
@vil02 ready for review

**Comment by vil02** [comment]
@sozelfist could you take a look? Do you have some reasonable idea how to _refactor_ `stable_matching` into smaller pieces?

**Comment by sozelfist** [comment]
> @sozelfist could you take a look? Do you have some reasonable idea of how to _refactor_ `stable_matching` into smaller pieces?

Sure, I will propose decomposing the `stable_matching` function into helper functions that handle the separated logic and make the main function more streamlined and focused.

File: src/greedy/stable_matching.rs [hunk 4]
```diff
@@ -0,0 +1,199 @@
+use std::collections::{HashMap, VecDeque};
+
+pub fn stable_matching(
```
Comment (sozelfist) [inline src/greedy/stable_matching.rs:3]: First of all, I will propose the way we can decompose the `stable_matching` function into helper functions along with their functionality. Based on the logic of the function, we can split it and place these small chunks into these helper functions.

- `initialize_men`: Initializes men’s proposal data.
- `initialize_women`: Initializes women’s partnership statuses.
- `precompute_woman_ranks`: Precomputes rankings of men for each woman.
- `process_proposal`: Processes a man’s proposal to a woman.
- `woman_prefers_new_man`: Determines if a woman prefers a new proposal over her current engagement.
- `engage_man`: Updates the engagement between a man and a woman.
- `finalize_matches`: Collects and returns the final stable matches.

File: src/greedy/stable_matching.rs [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+
+    let mut stable_matches: HashMap<String, String> = HashMap::new();
+    for (man, woman_option) in man_engaged {
+        if let Some(woman) = woman_option {
+            stable_matches.insert(man, woman);
+        }
+    }
+
+    stable_matches
+}
```
Comment (sozelfist) [inline src/greedy/stable_matching.rs:61]: The function logic could be like this
```rs
pub fn stable_matching(
    men_preferences: &HashMap<String, Vec<String>>,
    women_preferences: &HashMap<String, Vec<String>>,
) -> HashMap<String, String> {
    let (mut free_men, mut next_proposal) = initialize_men(men_preferences);
    let mut current_partner = initialize_women(women_preferences);
    let mut man_engaged = HashMap::new();

    let woman_ranks = precompute_woman_ranks(women_preferences);

    while let Some(man) = free_men.pop_front() {
        process_proposal(
            &man,
            &mut free_men,
            &mut current_partner,
            &mut man_engaged,
            &mut next_proposal,
            men_preferences,
            &woman_ranks,
        );
    }

    finalize_matches(man_engaged)
}
```

Comment (sozelfist) [reply src/greedy/stable_matching.rs:61]: Here are the helper functions implementation (just place code in right place):

```rs
fn initialize_men(men_preferences: &HashMap<String, Vec<String>>) -> (VecDeque<String>, HashMap<String, usize>) {
    let mut free_men = VecDeque::new();
    let mut next_proposal = HashMap::new();

    for man in men_preferences.keys() {
        free_men.push_back(man.clone());
        next_proposal.insert(man.clone(), 0);
    }

    (free_men, next_proposal)
}

fn initialize_women(women_preferences: &HashMap<String, Vec<String>>) -> HashMap<String, Option<String>> {
    let mut current_partner = HashMap::new();
    for woman in women_preferences.keys() {
        current_partner.insert(woman.clone(), None);
    }
    current_partner
}

fn precompute_woman_ranks(women_preferences: &HashMap<String, Vec<String>>) -> HashMap<String, HashMap<String, usize>> {
    let mut woman_ranks = HashMap::new();
    for (woman, preferences) in women_preferences {
        let mut rank_map = HashMap::new();
        for (rank, man) in preferences.iter().enumerate() {
            rank_map.insert(man.clone(), rank);
        }
        woman_ranks.insert(woman.clone(), rank_map);
    }
    woman_ranks
}

fn process_proposal(
    man: &str,
    free_men: &mut VecDeque<String>,
    current_partner: &mut HashMap<String, Option<String>>,
    man_engaged: &mut HashMap<String, Option<String>>,
    next_proposal: &mut HashMap<String, usize>,
    men_preferences: &HashMap<String, Vec<String>>,
    woman_ranks: &HashMap<String, HashMap<String, usize>>,
) {
    let man_pref_list = &men_preferences[man];
    let next_woman_idx = next_proposal[man];
    let woman = &man_pref_list[next_woman_idx];

    // Update man's next proposal index
    next_proposal.insert(man.to_string(), next_woman_idx + 1);

    if let Some(current_man) = current_partner[woman].clone() {
        // Woman is currently engaged, check if she prefers the new man
        if woman_prefers_new_man(woman, man, &current_man, woman_ranks) {
            engage_man(man, woman, free_men, current_partner, man_engaged, Some(current_man));
        } else {
            // Woman rejects the proposal, so the man remains free
            free_men.push_back(man.to_string());
        }
    } else {
        // Woman is not engaged, so engage her with this man
        engage_man(man, woman, free_men, current_partner, man_engaged, None);
    }
}

fn woman_prefers_new_man(
    woman: &str,
    man1: &str,
    man2: &str,
    woman_ranks: &HashMap<String, HashMap<String, usize>>,
) -> bool {
    let ranks = &woman_ranks[woman];
    ranks[man1] < ranks[man2]
}

fn engage_man(
    man: &str,
    woman: &str,
    free_men: &mut VecDeque<String>,
    current_partner: &mut HashMap<String, Option<String>>,
    man_engaged: &mut HashMap<String, Option<String>>,
    current_man: Option<String>,
) {
    man_engaged.insert(man.to_string(), Some(woman.to_string()));
    current_partner.insert(woman.to_string(), Some(man.to_string()));

    if let Some(current_man) = current_man {
        // The current man is now free
        free_men.push_back(current_man);
    }
}

fn finalize_matches(man_engaged: HashMap<String, Option<String>>) -> HashMap<String, String> {
    let mut stable_matches = HashMap::new();
    for (man, woman_option) in man_engaged {
        if let Some(woman) = woman_option {
            stable_matches.insert(man, woman);
        }
    }
    stable_matches
}
```

**Comment by sozelfist** [comment]
What do you think about the refactor, @vil02 and @BKarthik7?

**Comment by BKarthik7** [comment]
> What do you think about the refactor, @vil02 and @BKarthik7? 

I understood the refactor and is nice but if a newbie wants to understand the code i thought the original version was more readable it's just my opinion I am open to hear your point of view too.

**Comment by sozelfist** [comment]
The original version might be easier for beginners due to its linear structure, but the refactored version is better in terms of maintainability and best practices.

**Comment by BKarthik7** [comment]
@sozelfist 
done is there any thing else

**Comment by sozelfist** [comment]
I just proposed the refactored version of your code. Maybe @vil02 would have something to propose.

**Comment by sozelfist** [comment]
> Btw, why don't we write tests using macros? It makes the test more compact and we can suggest adding some edge tests.

You can consider this one for tests

**Review by vil02 — APPROVED** [review]
I think we can leave it as it is.

**Pull request merged**